### DISCUSSION TITLE
Questoin about tx-pps

### DISCUSSION THREAD

**Discussion by winnie022** [baidu/dperf#498] [Q&A]
Hello, 

I am new to dperf. Thank you for maintaining great tool.

1. I would like to test 1 MPPS, 2 MPPS, etc by gradually increasing the rate.
Is it possible to control pps rate on TX side?
I am currently explore flood under performance/tx-pps and seems like cc can control it.
There are multiple parameters like cc, cps, and keepalive, slow_start.
Even udp has slow start and keepalive features?
Is tx-pps the best option to test it?
I am using dperf on TX side and testpmd on RX side. (one direction)

2. If I can measure round-trip pps (bi-directional) to check packet loss, what is the best option in dperf and which one should I take a look under test directory?

3. Are there any docs or articles to understand these parameters in details?

Thanks,

**Comment by pengjianzhang** [comment]
> Hello,
> 
> I am new to dperf. Thank you for maintaining great tool.
> 
> 1. I would like to test 1 MPPS, 2 MPPS, etc by gradually increasing the rate.
>    Is it possible to control pps rate on TX side?

      yes. pps = cc *  ( 1s / keepalive-interval)

>    I am currently explore flood under performance/tx-pps and seems like cc can control it.
>    There are multiple parameters like cc, cps, and keepalive, slow_start.
>    Even udp has slow start and keepalive features?

       see more about keepalive  https://dperf.org/doc/html/configuration#section5

>    Is tx-pps the best option to test it?

      use 'cc', 'keepalive', etc to control it.

>    I am using dperf on TX side and testpmd on RX side. (one direction)

      please use dperf at both side. 
      if you only test tx-pps,  you should use 'flood' in client.conf, and server.conf use a diffrent 'listen port', so dperf server will drop all the recving packets from dperf client.


> 2. If I can measure round-trip pps (bi-directional) to check packet loss, what is the best option in dperf and which one should I take a look under test directory?

      dperf.org

> 3. Are there any docs or articles to understand these parameters in details?

   dperf.org

> 
> Thanks,

**Comment by winnie022** [comment]
Hello, 
Thank you for detailed response. I will check your references.


> please use dperf at both side. 
if you only test tx-pps,  you should use 'flood' in client.conf, and server.conf use a diffrent 'listen port', so dperf server will drop all the recving packets from dperf client.

I want to check the RX packets as well. So, I used testpmd. Do I know why you suggest using dperf in both side?
Also, in server side, how to configure config files to scale the server? In client, if I use more cpus, it uses multiple tx queues. In server side, does it the same? 

I have few more questions

1. I saw http example under test dir, does dperf have iperf-like tests without http requests? Also, does it compatible with those tools (iperf, etc)? 

2. There are udp under test dir, what's the main difference between udp client under test dir and flood example?
If I use client and server under test dir, does it scale by adding the cpus as well?

Thanks

**Reply by pengjianzhang** [reply]
> Do I know why you suggest using dperf in both side? 

dperf is desiged to work this way, dperf-client and dperf-server run together.  

> Also, in server side, how to configure config files to scale the server? 

https://dperf.org/


> In client, if I use more cpus, it uses multiple tx queues. In server side, does it the same?

yes.


> 1. I saw http example under test dir, does dperf have iperf-like tests without http requests? 

sorry, I don't get your question.

> Also, does it compatible with those tools (iperf, etc)?

No

> 2. There are udp under test dir, what's the main difference between udp client under test dir and flood example?

'flood': client just send packets out,  don't expect any response.

>    If I use client and server under test dir, does it scale by adding the cpus as well?

yes.


if you like dperf, could you please star it. thanks.

**Comment by winnie022** [comment]
> I saw http example under test dir, does dperf have iperf-like tests without http requests?
sorry, I don't get your question.

dperf always has http request as tcp payload? Are there option only simple tcp packets with dummy byte payload which is not http?

> 'flood': client just send packets out, don't expect any response.
I see. The other udp examples get response from server? I think this can be only tcp or http example.

> if you like dperf, could you please star it. thanks.
Sure! I did it!

Thanks

**Reply by pengjianzhang** [reply]
> dperf always has http request as tcp payload?

yes

>  Are there option only simple tcp packets with dummy byte payload which is not http?

payload_file Path


by the way, How did you know about dperf? Do your colleagues know dperf?

**Comment by winnie022** [comment]
Thank you for response.

I googled tools for traffic generator. It is really useful.
I will get back to you if I have more questions as I tried more.
